lib/ts/recipe/session/with-jwt/recipeImplementation.ts

```diff
@@ -0,0 +1,182 @@
+/* Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
+ *
+ * This software is licensed under the Apache License, Version 2.0 (the
+ * "License") as published by the Apache Software Foundation.
+ *
+ * You may not use this file except in compliance with the License. You may
+ * obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+import * as JsonWebToken from "jsonwebtoken";
+
+import { RecipeInterface } from "../";
+import { NormalisedAppinfo } from "../../../types";
+import { RecipeInterface as JWTRecipeInterface } from "../../jwt/types";
+import { SessionContainerInterface, TypeNormalisedInput, VerifySessionOptions } from "../types";
+import SessionClassWithJWT from "./sessionClass";
+
+export default function (
+    originalImplementation: RecipeInterface,
+    jwtRecipeImplementation: JWTRecipeInterface,
+    config: TypeNormalisedInput,
+    appInfo: NormalisedAppinfo
+): RecipeInterface {
+    // Time difference between JWT expiry and access token expiry (JWT expiry = access token expiry + EXPIRY_OFFSET_SECONDS)
+    const EXPIRY_OFFSET_SECONDS = 30;
+
+    function getJWTExpiry(accessTokenExpiry: number): number {
+        return accessTokenExpiry + EXPIRY_OFFSET_SECONDS;
+    }
+
+    return {
+        ...originalImplementation,
+        createNewSession: async function ({
+            res,
+            userId,
+            accessTokenPayload,
+            sessionData,
+        }: {
+            res: any;
+            userId: string;
+            accessTokenPayload?: any;
+            sessionData?: any;
+        }): Promise<SessionContainerInterface> {
+            let accessTokenValidityInSeconds = Math.ceil((await this.getAccessTokenLifeTimeMS()) / 1000);
+            let jwtResponse = await jwtRecipeImplementation.createJWT({
+                payload: {
+                    /* 
+                        We add our claims before the user provided ones so that if they use the same claims
+                        then the final payload will use the values they provide
+                    */
+                    sub: userId,
+                    iss: appInfo.apiDomain.getAsStringDangerous(),
```

Comment:
You are adding these extra claims only to the JWT payload. This means, that the access token payload doesn't have these. This means that when you refresh the session, they will go away?

Reply:
If this is true, and you hadn't already detected this in testing, it means you need to think harder about tests.

Reply:
Testing was in progress and this logic did break when I added the relevant tests. Fixed now